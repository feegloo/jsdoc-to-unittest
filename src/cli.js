import fs from 'fs';
import path from 'path';
import util from 'util';
import { argv } from 'yargs';
import print from './creator';

const readFileAsync = util.promisify(fs.readFile);

const basePath = path.resolve(__dirname, '..');

const fields = [
  'inline'
];

async function getConfig(input, output) {
  let configPath = '';
  const config = { // default values
    export: true,
  };
  if (argv.config) {
    configPath = path.resolve(basePath, argv.config, '..');
    Object.assign(config, JSON.parse(
      await readFileAsync(path.resolve(basePath, argv.config), 'utf-8')
    ));
  }

  fields.forEach((field) => {
    if (field in argv) {
      config[field] = argv[field];
    }
  });

  Object.assign(config, {
    input,
    output,
  });

  if (argv.header) {
    config.header = await readFileAsync(path.resolve(basePath, config.header), 'utf-8');
  } else if (config.header) {
    config.header = await readFileAsync(path.resolve(configPath, config.header), 'utf-8');
  }
  if (argv.footer) {
    config.footer = await readFileAsync(path.resolve(basePath, config.footer), 'utf-8');
  } else if (config.footer) {
    config.footer = await readFileAsync(path.resolve(configPath, config.footer), 'utf-8');
  }

  return config;
}

export default async (input = argv._[0], output = argv._[1]) => {
  const config = await getConfig(input, output);
  return print({
    ...config,
    get stdout() {
      return output === undefined || String(output).trim().length === 0;
    },
  });
};
