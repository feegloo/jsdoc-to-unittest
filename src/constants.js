export default {
  jest: {
    method: 'test',
  },

  karma: {
    method: 'it',
  },
};

export const tokens = {
  CALL: 'call',
  ACCESS: 'access',
  GET: 'get',
};
