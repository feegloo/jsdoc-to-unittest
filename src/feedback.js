export default (value) => {
  const obj = {
    value,
    type: typeof value,
  };

  return obj;
};
