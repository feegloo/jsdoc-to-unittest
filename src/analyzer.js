import { evaluate, evaluateAsync, getFunctionBody, getFunctionParams } from './utils';
import feedback from './feedback';
import joinPath from './path';

function collectFeedback(argumentsList) {
  return argumentsList.map(feedback);
}

function intercept(ret, _eval, { keys = {}, args = {} } = {}) {
  const obj = {
    path: [],
    access: [],
    calls: 0,
  };
  let current = null;
  let included = false;
  let excluded = false;
  const instance = new Proxy(function () {}, { // eslint-disable-line prefer-arrow-callback
    apply(target, thisArg, argumentsList) {
      if (argumentsList.some(item => item === instance)) {
        ret.pop();
        current = ret[ret.length - 1];
      }

      current.calls += 1;
      const argsList = new Array(argumentsList.length).fill(feedback(undefined));
      current.access.push('call', argsList);
      try {
        collectFeedback(JSON.parse(JSON.stringify(argumentsList))).forEach((item, i) => {
          argsList[i] = item;
        });
      } catch (ex) {}

      try {
        const ref = _eval(joinPath(current.path));
        if (typeof ref === 'function') {
          const args = {};
          getFunctionParams(ref).forEach((name, i) => {
            args[name] = argumentsList[i];
          });
          ret.push(...parseKey(getFunctionBody(ref), _eval, { args }));
        }
      } catch (ex) {}
      return instance;
    },
    construct(target, argumentsList) {
      if (argumentsList.some(item => item === instance)) {
        ret.pop();
        current = ret[ret.length - 1];
      }
      return instance;
    },
    has(target, key) {
      if (!Reflect.has(args, key)) {
        current = JSON.parse(JSON.stringify(obj));
        ret.push(current);
        included = false;
        excluded = false;
      } else {
        excluded = true;
      }
      return true;
    },
    get(target, key) {
      switch (key) {
        case Symbol.unscopables:
          return undefined;
        case Symbol.toPrimitive:
          return function (hint) {
            target.apply(this, arguments); // eslint-disable-line prefer-rest-params
            switch (hint) {
              case 'number':
                return 0;
              case 'string':
                return '';
              default:
                return true;
            }
          };
        case 'toJSON':
          return () => null;
        default:
          if (!excluded) {
            if (Object.keys(keys).includes(key)) {
              current.special = keys[key];
            }

            current.path.push(key);
            if (!included) {
              included = true;
              current.access.push('access');
            } else {
              current.access.push('get');
            }
          }
      }

      if (args[key]) {
        return args[key];
      }

      return instance;
    },
  });

  return instance;
}

export function parseKey(exp, _eval = () => {}, { keys = {}, args = {} } = {}) {
  const ret = [];
  evaluate(`with(s){\n${exp}\n}`, {
    s: intercept(ret, _eval, { keys, args }),
  });
  return ret;
}

export async function parseKeyAsync(exp, _eval) {
  const ret = [];
  await evaluateAsync(`with(s){\n${exp}\n}`, {
    s: intercept(ret, _eval, {
      follow: true,
      keys: {
        then: 'Promise',
        catch: 'Promise',
      },
    }),
  });

  return ret.map((item) => {
    if (item.special !== 'Promise') {
      return item;
    }

    const { path, access } = item;
    if (path[0] === '_asyncTestExports') {
      path.shift();
      access.splice(1, 1);
    }

    const firstThenIndex = path.indexOf('then');
    path.splice(firstThenIndex, path.length);
    const mergedAccess = access.reduce((acc, cur) => {
      if (cur === 'access') {
        acc.push([cur]);
      } else if (cur === 'get') {
        acc.push([cur]);
      } else {
        acc[acc.length - 1].push(cur);
      }

      return acc;
    }, []);
    mergedAccess.splice(firstThenIndex, mergedAccess.length);
    const concat = mergedAccess.reduce((acc, cur) => [...acc, ...cur], []);

    return {
      ...item,
      access: concat,
      calls: concat.reduce((acc, cur) => {
        if (cur === 'call') {
          return acc + 1;
        }

        return acc;
      }, 0),
      path,
    };
  });
}

export function getPath(func, follow) {
  return parseKey(func, this, { follow }).reduce((acc, { path }) => [...acc, path], []);
}

export function isCallable(func, toObserve, all = false) {
  return parseKey(func, this)
    .filter(({ path }) => toObserve.includes(path.join('.')))
    [all ? 'every' : 'some'](({ calls }) => calls > 0); // eslint-disable-line no-unexpected-multiline
}

function sandbox(func) {
  return Function(
    's, ...args',
    `s.args=args;with(s){return(${func.toString()}).apply(this,arguments)}`,
  );
}

export function fallToGlobal(target, _eval, oldKey = '') {
  const mappedArgs = new WeakMap();
  return new Proxy(target, {
    has: () => true,
    set(_, key, value) {
      if (key === 'args') {
        mappedArgs.set(target, value);
      }

      return true;
    },
    get(_, key) {
      if (key === '_asyncTestExports') { // fixme: that _asyncTestExports
        return fallToGlobal(target, _eval);
      }

      if (key === Symbol.unscopables) return null;
      if (key === 'arguments') {
        return mappedArgs.get(target) || [];
      }

      if (Reflect.has(target, key)) {
        const newTarget = target[key];
        if (typeof newTarget !== 'object' && typeof newTarget !== 'function') {
          return newTarget;
        }

        return fallToGlobal(newTarget, _eval, `${oldKey}.${key}`);
      }

      try {
        let func;
        if (oldKey) {
          func = _eval(`${oldKey}.${key}`.replace('.', ''));
        } else {
          func = _eval(key);
        }

        if (typeof func === 'function') {
          return sandbox(func).bind(null, fallToGlobal(target, _eval, `${oldKey}.${key})`));
        }

        return func;
      } catch (ex) {}
    },
  });
}

function filterAccesses(parsed, filter) {
  if (filter.length) {
    const paths = filter.reduce((acc, path) => {
      acc.push(...getPath(path).map(item => item.join('.')));
      return acc;
    }, []);

    parsed = parsed.filter(({ path }) => paths.includes(path.join('.')));
  }

  return parsed.reduce((acc, { access }) => [...acc, access], []);
}

export function listAccesses(code, filter = []) {
  try {
    return filterAccesses(parseKey(code, this), filter);
  } catch (ex) {
    return [];
  }
}

export async function listAccessesAsync(code, filter = []) {
  try {
    return filterAccesses(await parseKeyAsync(code, this), filter);
  } catch (ex) {
    return [];
  }
}
