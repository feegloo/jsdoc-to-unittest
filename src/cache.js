import * as acorn from 'acorn';
import { isPrimitive } from './utils';

export default {
  ast: new class extends WeakMap { // eslint-disable-line new-parens
    get(target) {
      if (isPrimitive(target)) {
        return acorn.parse(target);
      }

      let ast = super.get(target); // eslint-disable-line no-shadow
      if (ast !== undefined) {
        return ast;
      }

      ast = acorn.parse(target);
      this.set(target, ast);
      return ast;
    }
  },

  analyzer: new class extends Map {
    cache(key, result) {

    }
  },
};
