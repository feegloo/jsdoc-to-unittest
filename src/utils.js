import escodegen from 'escodegen';
import * as acorn from 'acorn';
import * as walk from 'acorn/dist/walk';
import cache from './cache';

export function evaluate(code, args = {}, strict = false) {
  return Function(Object.keys(args).join(','), `${strict ? '\'use strict\';' : ''}${code}`)(...Object.values(args));
}

export const AsyncFunction = (async () => {}).constructor;

export async function evaluateAsync(code, args = {}, strict = false) {
  return AsyncFunction(Object.keys(args).join(','), `${strict ? '\'use strict\';' : ''}${code}`)(...Object.values(args));
}

export const isAsyncFunction = func => typeof func === 'function' && func instanceof AsyncFunction;

export function validateSyntax(code) {
  try {
    evaluate(code, {}, true);
  } catch (ex) {
    if (ex instanceof SyntaxError) {
      return {
        ex,
      };
    }
  }

  return {
    ex: null,
  };
}

export function isFunction(str, scope = {}) {
  try {
    return typeof evaluate(`return (${str})`, scope) === 'function';
  } catch (ex) {}

  return false;
}

const semicolons = /^[\s;]+|[\s*;]+$/g;

export const stripSemicolons = str => str.replace(semicolons, '');

export const stripComments = (str) => {
  try {
    return escodegen.generate(acorn.parse(str), {
      format: {
        indent: {
          style: '  ',
          base: 0,
          adjustMultilineComment: false,
        },
      },
    });
  } catch (ex) {
    return str;
  }
};

const evalFunc = 'func => eval(func)';

export function wrap(src, mocks, mockName = 'mock') {
  if (typeof src !== 'string') {
    return '';
  }

  if (validateSyntax(src).ex !== null) {
    return '';
  }

  const stripped = stripComments(src);
  const strippedSemicolons = stripSemicolons(stripped);

  if (!strippedSemicolons.length) {
    return '';
  }

  if (isFunction(strippedSemicolons, this)) {
    return mocks ? `${mockName}(${strippedSemicolons}, ${mocks}, ${evalFunc})` : strippedSemicolons;
  }

  const ast = cache.ast.get(src);
  if (ast.body.length === 1 &&
    ast.body[0].type === 'ExpressionStatement' &&
    ast.body[0].expression.type === 'CallExpression'
  ) {
    const strippedSrc = stripSemicolons(stripComments(src));
    return mocks ? `${mockName}(() => ${strippedSrc}, ${mocks}, ${evalFunc})` : strippedSrc;
  }

  return mocks ? `${mockName}(${src}, ${evalFunc}), ${mocks}, ${evalFunc})` : `evaluate(() => { ${stripComments(src)} }, ${evalFunc})`;
}

export const isPrimitive = sth => sth === null || (typeof sth !== 'object' && typeof sth !== 'function');

export const assertAccess = target => new Proxy(() => {}, {
  has: () => true,
  get(_, key) {
    if (key === Symbol.unscopables) return null;

    if (!isPrimitive(target) && Reflect.has(target, key)) {
      const value = target[key];
      if (isPrimitive(value)) {
        return value;
      }

      if (typeof value === 'function') {
        return function () {
          try {
            // eslint-disable-next-line prefer-rest-params
            return assertAccess(value.apply(this, arguments));
          } catch (ex) {}
        };
      }

      return assertAccess(value);
    }

    return assertAccess({});
  },
});

export function getFunctionBody(func) {
  const nodes = [];
  walk.simple(cache.ast.get(func), {
    FunctionDeclaration(node) {
      nodes.push(node);
    },
    ArrowFunctionExpression(node) {
      nodes.push(node);
    },
  });

  if (nodes.length && nodes[0].body !== undefined) {
    return escodegen.generate(nodes[0].body);
  }

  if (typeof func === 'function') {
    return Reflect.apply(Function.toString, func, []);
  }

  return func;
}

export function getFunctionParams(func) {
  const nodes = [];
  walk.simple(cache.ast.get(func), {
    FunctionDeclaration(node) {
      nodes.push(node);
    },
  });

  if (nodes.length && nodes[0].params !== undefined) {
    return nodes[0].params.map(item => item.name);
  }

  return [];
}
