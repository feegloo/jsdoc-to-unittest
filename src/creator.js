import prettify from './prettifier';
import { wrap, validateSyntax } from './utils';
import { readFile, writeFile, toStdout } from './fs';

class TestItem {
  constructor({
    async,
    name,
    code,
    type,
    result,
    imports,
    mocks,
  }) {
    this.async = async;
    this.name = name;
    this.code = code;
    this.type = type;
    this.result = result;
    this.imports = imports;
    this.mocks = JSON.stringify(mocks);
    this.mockName = async ? 'mock.async' : 'mock';
    this.valid = false;
    this.output = this.print();
  }

  printName(code) {
    return `test('${this.name}', ${this.async ? 'async ' : ''}() => {
      ${code}
    })`;
  }

  renderEquality() {
    if (this.type !== 'no-throw' && typeof this.result === 'string' && !this.result.length) {
      this.type = 'no-throw';
      return this.renderEquality();
    }

    switch (this.type) {
      case 'instance':
        return `toBeOneInstanceOf(${JSON.stringify(this.result)}); // fixme: could be replaced with something more specific`;
      case 'value':
        return `toBe(${this.result});`;
      case 'no-throw':
        return 'not.toThrow(); // fixme: could be replaced with something more specific';
      default:
        return `toEqual(${this.result});`;
    }
  }

  print() {
    if (validateSyntax(this.renderEquality()).ex !== null) {
      this.type = 'no-throw';
    }

    const wrapped = wrap(this.code, this.mocks, this.mockName);
    const code = `expect(${this.async ? 'await ' : ''}${wrapped}).${this.renderEquality()}`;
    const test = this.printName(code);

    if (!wrapped.length) {
      return this.printError({ ex: new SyntaxError('Empty expect'), code });
    }

    const { ex } = validateSyntax(test);
    if (ex) {
      return this.printError({ ex, code });
    }

    this.valid = true;
    return test;
  }

  [Symbol.toPrimitive](hint) {
    switch (hint) {
      case 'string':
        return this.output;
      /* istanbul ignore next */
      default:
        return Object.prototype.toString.call(this);
    }
  }

  printError({ ex, code }) {
    return this.printName(`throw new ${ex.constructor.name}('${ex.message}');\n/*${code}*/`);
  }
}

class Test {
  constructor({
    filename = '',
    stdout,
    samples,
    imports,
    exports,
    async,
    partials = [],
  }) {
    this.filename = filename;
    const splitFilename = filename.split('/');
    this.exportsFileName = splitFilename[splitFilename.length - 1].replace(/\.js$/, '.exports.js');
    this.stdout = stdout;
    this.samples = samples;
    this.imports = imports;
    this.exports = exports;
    this.partials = partials;
    this.async = async;
    this.stats = {
      total: 0,
      valid: 0,
    };
  }

  printImports() {
    if (this.stdout) return '';
    return `import __imports__, {
       ${this.imports.join(',\n')}
      } from './${this.exportsFileName}';
      
      global.__imports__ = __imports__;
      Object.assign(global, __imports__);
      `;
  }

  printPartials() {
    return Promise.all(this.partials.map(partial => readFile(partial)));
  }

  printExports() {
    if (!this.imports.length) {
      return '';
    }

    return prettify([
      this.exports.map(func => `export ${func}`).join('\n\n'),
      ';',
      `export default { ${this.imports.join(',\n')} };`,
    ].join('\n'));
  }

  printSample({ examples, name = (examples[0] || []).name }) {
    if (!examples.length) return '';
    const tests = examples.map((args, i) => {
      const item = new TestItem({
        async: this.async,
        imports: this.imports,
        ...args,
        name: `example ${i + 1}`,
      });
      this.stats.total += 1;
      if (item.valid) {
        this.stats.valid += 1;
      }
      return String(item);
    }).join('\n\n');

    return `describe('${name}', ${this.async ? 'async ' : ''}() => {
      ${tests}
    });`;
  }

  printStats() {
    if (this.stats.valid === 0) {
      return '// Valid tests: 0%';
    }

    return `// Valid tests: ${((this.stats.valid / this.stats.total) * 100).toFixed(2)}%`;
  }

  async print() {
    const samples = this.samples.map(this.printSample, this).join('\n');
    return prettify([
      this.printStats(),
      this.printImports(),
      ...(await this.printPartials()),
      samples,
    ].join('\n'));
  }

  async write() {
    let output = '';
    try {
      output = await this.print();
    } catch (ex) {
      output = JSON.stringify(ex.message.toString());
    }

    if (this.stdout) {
      return toStdout(prettify(`${this.printExports()}\n${output}`));
    }

    return Promise.all([
      writeFile(this.exportsFileName, this.printExports()),
      writeFile(this.filename, output),
    ]);
  }
}

export default async (args) => {
  const test = new Test(args);
  // todo: move (from cli) readFile here etc...
  return test.write();
};
