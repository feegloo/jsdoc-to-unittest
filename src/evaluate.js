import { getFunctionBody } from './utils';

export default (src, _eval) => _eval(getFunctionBody(src));

