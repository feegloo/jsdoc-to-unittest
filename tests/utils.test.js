import { toResult, isFunction, wrap, assertAccess, stripComments, evaluate, isPrimitive } from 'src/utils';

function call() { // eslint-disable-line no-unused-vars
  return '2';
}

function call2() { // eslint-disable-line no-unused-vars
  return 10;
}

const func = `
  () => {
    var arr = [0, 2];
    var x = call2();
    return arr[1] + +call() + x;
  }
`;

const func2 = `
  var arr = [0, 2];
  var x = call2();
  call() + x;
`;

const func3 = `
  // d
  //dd
  call(); // d
`;

function exec(src) {
  return eval(wrap(src));
}

describe('#wrap', () => {
  test('works', () => {
    expect(exec('call()')).toBe('2');
    expect(exec('call();')).toBe('2');
    expect(exec('2 + +call();')).toBe(4);
    expect(exec('call();call2();')()).toBe(10);
    expect(exec('call();\n\ncall2();')()).toBe(10);
    expect(exec('\n\n;call();call();\n\ncall2();')()).toBe(10);
    expect(exec(func)()).toBe(14);
    expect(exec(func2)()).toBe('210');
    expect(exec(func3)).toBe('2');
  });

  test('mock functions', () => {
    expect(wrap.bind({ call }, 'call', true)()).toBe('mock(call)');
    expect(wrap('call()', true)).toBe('mock(() => call())');
    expect(wrap('call()\ncall2()', true)).toBe('mock(() => {\n' +
      '        call();\n' +
      '        return (call2());\n' +
      '      })');
  });

  test('clever wrapping', () => {
    expect(wrap(`easy.utils.each([1, 2, 3], function (value, index) {
      easy.console.log(index, value);
    });`)).toBe('easy.utils.each([1, 2, 3], function (value, index) {      easy.console.log(index, value)    })');
  });

  test('never throws', () => {
    expect(wrap(false)).toBe('');
    expect(wrap('')).toBe('');
    expect(wrap('--')).toBe('');
    expect(wrap()).toBe('');
    expect(wrap(Function)).toBe('');
  });
});

describe('#stripComments', () => {
  expect(stripComments('test //yes')).toBe('test ');
  expect(stripComments('test / /yes')).toBe('test / /yes');
  expect(stripComments('test /////yes')).toBe('test ');
  expect(stripComments('test//yes')).toBe('test');
  expect(stripComments('test/* ddd */test')).toBe('testtest');
  expect(stripComments('test/*ddd*/test')).toBe('testtest');
  expect(stripComments('test/**/test')).toBe('testtest');
  expect(stripComments(`/**
   * Shared context (XDM RPC) object reference
   * @memberof easy.constants
   * @constant
   * @default
   */
   easy.bla.bla
   `).replace(/\n/g, '').trim()).toBe('easy.bla.bla');
});

describe('#assertAccess', () => {
  test('returns value', () => {
    expect(Function('s', 'with(s){ return foo }')(assertAccess({ foo: 2 }))).toBe(2);
    expect(Function('s', 'with(s){ return foo.bar.x }')(assertAccess({ foo: { bar: { x: '100' } } }))).toBe('100');
    expect(Function('s', 'with(s){ return foo.bar().x }')(assertAccess({ foo: { bar: () => ({ x: '100' }) } }))).toBe('100');
  });

  test('no ReferenceError', () => {
    expect(Function('s', 'with(s){ foo }')(assertAccess({}))).toBe(undefined);
    expect(Function('s', 'with(s){ foo || bar }')(assertAccess({}))).toBe(undefined);
    expect(() => Function('s', 'with(s){ foo }')({})).toThrow(ReferenceError);
  });

  test('no TypeError', () => {
    expect(Function('s', 'with(s){ foo.bar.xyz }')(assertAccess({}))).toBe(undefined);
    expect(() => Function('s', 'with(s){ foo.bar.xyz }')()).toThrow(TypeError);
    expect(() => Function('s', 'with(s){ foo.bar.xyz }')({ foo: {} })).toThrow(TypeError);
  });
});

describe('#evaluate', () => {
  test('evaluates expression', () => {
    expect(evaluate('2 + 2')).toBe(undefined);
  });

  test('sloppy by default', () => {
    expect(() => evaluate('with({}){}')).not.toThrow();
  });

  test('accepts arguments mode is supported', () => {
    expect(evaluate('return a + b', { a: 1, b: 5 }, true)).toBe(6);
  });

  test('strict mode is supported', () => {
    expect(() => evaluate('with({}){}', {}, true)).toThrow();
  });

  test('returns value', () => {
    expect(evaluate('return 2 + 2')).toBe(4);
  });
});

describe('#isPrimitive', () => {
  test('returns true for primitive type', () => {
    expect(isPrimitive('d')).toBe(true);
    expect(isPrimitive(2)).toBe(true);
    expect(isPrimitive(null)).toBe(true);
    expect(isPrimitive(undefined)).toBe(true);
    expect(isPrimitive()).toBe(true);
    expect(isPrimitive(Symbol('dd'))).toBe(true);
    expect(isPrimitive(true)).toBe(true);
  });

  test('returns false for non-primitive type', () => {
    /* eslint-disable no-new-wrappers */
    expect(isPrimitive(new String('d'))).toBe(false);
    expect(isPrimitive(new Number(2))).toBe(false);
    expect(isPrimitive(new Boolean(2))).toBe(false);
    /* eslint-enable no-new-wrappers */
    expect(isPrimitive({})).toBe(false);
    expect(isPrimitive(() => {})).toBe(false);
  });
});

describe('#isFunction', () => {
  expect(isFunction('[].forEach')).toBe(true);
});

describe('#toResult', () => {
  expect(toResult(';[].forEach')).toBe('[].forEach');
  expect(toResult(';[].forEach;;')).toBe('[].forEach');
  expect(toResult('[].forEach')).toBe('[].forEach');
  expect(toResult(';;[].forEach')).toBe('[].forEach');
  expect(toResult(';;[].forEach;;')).toBe('[].forEach');
});
