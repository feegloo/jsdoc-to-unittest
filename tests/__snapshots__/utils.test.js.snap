// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getFunctionBody arrow functions 1`] = `
"{
    const arr = [
        10,
        5,
        3
    ];
    arr[2];
}"
`;
